### ISSUE TITLE
Makumba support for full-text search in non-public sites

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#535]
**Reported by @rudolfmayer on 21 Sep 2007 21:57 UTC**
MySQL supports full-text search by a specific match(..) agains (..) syntax (http://dev.mysql.com/doc/refman/5.0/en/fulltext-search.html). 

Also, MS SQL has something similar (http://msdn2.microsoft.com/en-us/library/ms142547.aspx)

The fundamental difference to having a simple LIKE %...% query is that
- it allows for not exact matches, query expansion, boolean matching
- it gives a score and thereby rank to the results, which allows to present most relevant matches first

It would be good for Makumba to provide some support to use these features:

- create fulltext indices, if the database allows that. indices could be created once makumba observes a page using full text search on a specific field. Alternatively, the indices could be defined manually per field in the MDD e.g.

- support fulltext queries using a specific syntax. seems OQL does not define this, neither does SQL, so if we do it, we need to find something new or follow MySQL or MS SQL.

- sorting the results should be doable in a list

- the score should be available to mak:value in some way.
 
i just saw now that some ideas were also in bug 258, but that was specific on PA so i will close it

Migrated-From: http://trac.makumba.org/ticket/987

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 22 Sep 2007 10:31 UTC**
anything in some OQL or HQL that resembles this?

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 22 Sep 2007 18:43 UTC**
(In reply to comment #1)
> anything in some OQL or HQL that resembles this?


not from what i have seen, neither in SQL. 

some links, besides the one in comment 0, that might be interesting:

- small overview: http://www.searchtools.com/info/database-search.html

- Postgresql: http://www.sai.msu.su/~megera/postgres/fts/doc/

**Comment by ghost** [comment]
**Comment by marius.andra@best.eu.org on 12 Oct 2007 22:51 UTC**
MySQL fulltext search is interesting. I've had a lot of trouble and fun with it.

First up, there are two modes to MySQL fulltext search: 
- Boolean mode
- Regular mode

Boolean mode is the mode where a text string either exists or not, which means you can compose cool contains-all queries. The other, regular, mode computes some "match" for every row and (if specified) sorts them by the highest. To my experience this has so far yielded worse results.

The other problem is that MySQL fulltext boolean search needs at query string of at least 4 characters to work. This is a pain if searching for example for "GA 2007"... it would only search for "2007"... Which means that that for smaller strings we'd still need to use LIKE queries, etc. 


So in conclusion, mysql fulltext search sucks. It's better than nothing, but are there other alternatives/tools/software that could be used?


Other than that, I could try doing something to implement MySQL fulltext searches.

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 19 Nov 2007 09:59 UTC**
hey!

(In reply to comment #3)
> Boolean mode is the mode where a text string either exists or not, which means
> you can compose cool contains-all queries. The other, regular, mode computes
> some "match" for every row and (if specified) sorts them by the highest. To my
> experience this has so far yielded worse results.


worse in what was found, or simply the ranking result not being good?

> The other problem is that MySQL fulltext boolean search needs at query string
> of at least 4 characters to work. This is a pain if searching for example for
> "GA 2007"... it would only search for "2007"... Which means that that for
> smaller strings we'd still need to use LIKE queries, etc. 


that kind of sucks..

> So in conclusion, mysql fulltext search sucks. It's better than nothing, but
> are there other alternatives/tools/software that could be used?


there is other software around, for sure. but from my experiences with playing around with lucene, anything that is outside the database is a pain on the jsp programme side to merge results from DB (to limit results for e.g. access rights) and lucene for the actual search.

something interesting could be http://sphinxsearch.com/index.html, which kind of plugs into mysql.

> Other than that, I could try doing something to implement MySQL fulltext
> searches.


basically it would be
- find a syntax we want to use in OQL, and make the anaylser work with it
- translate that to the SQL specific queries (e.g. mysql and postgres use different things)
- that's it, i think

problem here - mysql supports full text ONLY with myisam, but we use inno db for transaction support... and you can't use the query w/o the index:

mysql> SELECT * FROM general_archive_File__convertedContent_  WHERE MATCH (text_) AGAINST ('itc');
ERROR 1214 (HY000): The used table type doesn't support FULLTEXT indexes


and trying to create it obviously fails:
mysql> create fulltext index abc ON  general_archive_File__convertedContent_ (text_);
ERROR 1214 (HY000): The used table type doesn't support FULLTEXT indexes

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 23 Mar 2008 23:52 UTC**
Rudi tells me that there is an expectation from users that they will have a google-like interface which would search through all _pages_ (not objects) accessible to them. in that respect most comments on this bug (which talk about mysql data, not about pages) are on the wrong track...

the classical solution for public sites is crawling + indexing + displaying search results in "relevance" order, in page context. the last two seem to be provided by Lucene.

this is a very challenging problem computationally for sites which have multiple user roles, like Karamba. it is practically impossible to crawl+index for every user (since users see different content) so crawling for one imaginary user who is in no team seems to be one solution.

what strikes me is that even if it "knows" a lot about the site (data structures, page structure) makumba cannot help much in this process. one way makumba could help is bug 1033. but i really think we should look for more ways, and try to find generic Makumba solutions, not just Karamba-specific.

we discussed about a generic Makumba index which would tell in which table and in which row a certain word can be found. however, even after knowing the row, one would have to know in which page that field of that row is displayed (we know that from page analysis), and whether the user ever has access to that page. declarative per-object security may help this. still this was turned down by Rudi, I don't remember why :)

Rudi also mentioned a blog about all this, would be nice to link it here, if this bug entry is indeed suitable.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 4 Jul 2008 12:44 UTC**
Rudi please reply to comment 5 :)

at the recent Stockholm congress we came up with the idea that having a set of authorization functions (e.g. in MakumbaController) can help full text search over authorized pages because you can determine in which accounts to crawl so that the number of pages crawled is maximized, and the total number of users having at least those accounts' access rights is maximized. That way you can determine like 5 crawling sets

can you also define "crawling regions" so if user set A have all rights of user set B, I just crawl for the pages that A can access but B cannot, and for the rest I use the results of the B crawling?

the other idea: can we crawl for a super-user (i.e. all possible pages), and then simply filter the search results using the constraints in MakumbaController.properties?

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 4 Jul 2008 13:02 UTC**
(In reply to comment #6)
> the other idea: can we crawl for a super-user (i.e. all possible pages), and
> then simply filter the search results using the constraints in
> MakumbaController.properties?


yes, that would be the prefered one, and that is also what i implemented in my prototype. what i do is
- search the index
- set the pointers of the results as vector in the pageContext
- make a mak:list where i compose my access control query, and say "type in SET $resultsFromLucene". Additionally, a lot of other things are set, like for each doc the paths, the scores, the matching text snippets..

This is based on the assumption that a Lucene search is fast anyhow (which it is).

so, maybe one thing that Makumba can do here it to provide some integration with lucene to do some of those things automatically (i.e. setting attributes)

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 4 Jul 2008 16:12 UTC**
(In reply to comment #7)
> yes, that would be the prefered one, and that is also what i implemented in my
> prototype. what i do is
> - search the index
> - set the pointers of the results as vector in the pageContext
> - make a mak:list where i compose my access control query, and say "type in SET
> $resultsFromLucene". Additionally, a lot of other things are set, like for each
> doc the paths, the scores, the matching text snippets..
> 
> This is based on the assumption that a Lucene search is fast anyhow (which it
> is).
> 
> so, maybe one thing that Makumba can do here it to provide some integration
> with lucene to do some of those things automatically (i.e. setting attributes)
> 

well my idea is rather different:
- get a result set from lucene, in the form of URLs with parameters
- for each URL, check against MakumbaController.properties authorization constraints. remember this will also check wether the page is accessible with the given parameters!
- if any of them fails the authorization check, remove the URL from the result set
- show all other results

sounds too simple to be correct :)

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 4 Jul 2008 16:44 UTC**
(In reply to comment #8)
> sounds too simple to be correct :)

oh i realize the problem: performance. however, since search results are typically displayed up to 50/page, you will need to do up to 50 queries per result page. furthermore that can be reduced by clustering the resulting URLs so more of them match a pattern from MakumbaController.properties. Then you can check just once per pattern.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 4 Jul 2008 16:51 UTC**
(In reply to comment #9)

> oh i realize the problem: performance. however, since search results are
> typically displayed up to 50/page, you will need to do up to 50 queries per
> result page. furthermore that can be reduced by clustering the resulting URLs
> so more of them match a pattern from MakumbaController.properties. Then you can
> check just once per pattern.

oh, actually not always because of the URL parameters which may differ... however: 
- you can compute all non-parametric authorization patterns, so those will not generate any queries
- you can cache all patterns that did pass, so you don't run a query again for the next search result that passes.

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:22 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:36 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 14:15 UTC**
